### PR TITLE
arc: add XY mem support

### PR DESCRIPTION
**Pull request by SiyuanCheng-CN**
## Introduction
The XY Memory is a feature commonly found in DSP processors to increase the DSP performance. The XY component allows a ARC processor to implicitly load source operands and store results into a closely coupled memory using a single instruction. This results in dense code and high performance. 

## Proposed change
Add XY memory for ARC EM9D/EM11D processors including em_starterkit, em_starterkit_em11d. emsdp_em9d, nsim_em, iotdk.
Add XY memory into arc mpu region.

### CODE REVIEW

File: dts/bindings/arc/arc,xccm.yaml
Comment (galak) [inline dts/bindings/arc/arc,xccm.yaml:1]: copyright doesn't look correct, year, and name.

File: dts/bindings/arc/arc,yccm.yaml
Comment (galak) [inline dts/bindings/arc/arc,yccm.yaml:1]: copyright doesn't look correct, year, and name.

File: dts/bindings/arc/arc,yccm.yaml
Comment (galak) [inline dts/bindings/arc/arc,yccm.yaml:4]: Can you add a little more detail to description.

File: dts/bindings/arc/arc,xccm.yaml
Comment (galak) [inline dts/bindings/arc/arc,xccm.yaml:4]: Can you add a little more detail to description.

File: boards/arc/nsim/nsim.dtsi
Comment (galak) [inline boards/arc/nsim/nsim.dtsi:12]: Where are these ifdef defines getting defined?

File: boards/arc/nsim/nsim_em.dts
Comment (galak) [inline boards/arc/nsim/nsim_em.dts:11]: Where are these ifdef defines getting defined?

**Review by galak — CHANGES_REQUESTED** [review]
dts bindings need fixing and the #defines in .dtsi files don't seem correct in that where are those defines getting set?

File: dts/bindings/arc/arc,xccm.yaml
Comment (SiyuanCheng-CN) [reply dts/bindings/arc/arc,xccm.yaml]: Done, thanks for your check.

File: boards/arc/nsim/nsim_em.dts
Comment (SiyuanCheng-CN) [reply boards/arc/nsim/nsim_em.dts:11]: I followed the style of other memory, such as [ICCM and DCCM](https://github.com/zephyrproject-rtos/zephyr/blob/main/boards/arc/nsim/nsim-ccm-mem.dtsi#:~:text=%23ifndef%20ICCM_ADDR,%23endif). Use #ifdef is to have different definition for special application, just like [arc_sec_application](https://github.com/zephyrproject-rtos/zephyr/blob/main/samples/boards/arc_secure_services/nsim_sem_normal.dts#:~:text=%23define%20ICCM_ADDR%2040000,DCCM_SIZE%20DT_SIZE_K(256)).

**Review by ruuddw — DISMISSED** [review]

**Review by abrodkin — DISMISSED** [review]

**Comment by evgeniy-paltsev** [comment]
Hi @SiyuanCheng-CN ,

as I can see in one commit you are removing ICCM_ADDR, ICCM_SIZE, DCCM_ADDR, DCCM_SIZE definitions and in second commit you are bringing them back. I guess you need to remove that part from both commits.


PS: I'm talking about removing and adding that part:
```
#ifndef ICCM_ADDR
#define ICCM_ADDR 0
#endif

#ifndef ICCM_SIZE
#define ICCM_SIZE DT_SIZE_K(1024)
#endif

#ifndef DCCM_ADDR
#define DCCM_ADDR 80000000
#endif

#ifndef DCCM_SIZE
#define DCCM_SIZE DT_SIZE_K(1024)
#endif
```

**Comment by SiyuanCheng-CN** [comment]
> Hi @SiyuanCheng-CN ,
> 
> as I can see in one commit you are removing ICCM_ADDR, ICCM_SIZE, DCCM_ADDR, DCCM_SIZE definitions and in second commit you are bringing them back. I guess you need to remove that part from both commits.
> 
> PS: I'm talking about removing and adding that part in `boards/arc/nsim/nsim.dtsi`:
> 
> ```
> #ifndef ICCM_ADDR
> #define ICCM_ADDR 0
> #endif
> 
> #ifndef ICCM_SIZE
> #define ICCM_SIZE DT_SIZE_K(1024)
> #endif
> 
> #ifndef DCCM_ADDR
> #define DCCM_ADDR 80000000
> #endif
> 
> #ifndef DCCM_SIZE
> #define DCCM_SIZE DT_SIZE_K(1024)
> #endif
> ```

**Review by evgeniy-paltsev — DISMISSED** [review]

**Review by ruuddw — DISMISSED** [review]

File: boards/arc/nsim/nsim_em.dts
Comment (fabiobaltieri) [reply boards/arc/nsim/nsim_em.dts:11]: That does not feel like a correct use of dts preprocessor macro, why not delete the existing node and redefine it how it's supposed to be instead? That way you wyould not have to do quirky things like `/* must be before include of nsim_em.dtsi */`. Not saying it should be fixed in this PR but maybe this should avoid copying that pattern (and then that can be fixed separately).

Comment (SiyuanCheng-CN) [reply boards/arc/nsim/nsim_em.dts]: removed `#ifndef` to avoid mislead

**Review by abrodkin — DISMISSED** [review]

**Review by ruuddw — DISMISSED** [review]

**Comment by mbolivar-nordic** [comment]
We can't merge this with the merge commit; please do a clean rebase

**Review by ruuddw — DISMISSED** [review]

**Review by abrodkin — DISMISSED** [review]

**Review by evgeniy-paltsev — CHANGES_REQUESTED** [review]
Hi @SiyuanCheng-CN,

Le's squash `arc: optimize XY mem dts definition` commit into previous one as it's only fixup - so we don't want to have it as a separate commit.

And let's add a bit more info into the original `arc: add XY mem support` commit about what we do and why (at least please copy introduction section from this PR to commit message)

Thanks!

**Review by ruuddw — DISMISSED** [review]

File: boards/arc/nsim/nsim_em.dts
Comment (evgeniy-paltsev) [inline boards/arc/nsim/nsim_em.dts:28]: Given that we don't actually use XY on `nsim_em` platform and we add `nsim_em11d` in DSP PR for DSP and XY testing - I'd propose to avoid unnecessary `nsim_em` platform changes.

Same for boards/arc/nsim/support/nsim_em.props & boards/arc/nsim/support/mdb_em.args files.

**Review by evgeniy-paltsev — CHANGES_REQUESTED** [review]

File: include/zephyr/arch/arc/v2/linker.ld
Comment (evgeniy-paltsev) [inline include/zephyr/arch/arc/v2/linker.ld:67]:
```suggestion
	XCCM  (rw)  : ORIGIN = XCCM_START,  LENGTH = XCCM_SIZE
```

File: boards/arc/iotdk/arc_mpu_regions.c
Comment (ruuddw) [inline boards/arc/iotdk/arc_mpu_regions.c:27]: iotdk has an MPU that supports overlapping regions, I don't think REGION_DYNAMIC is required (DCCM doesn't have the attribute either).
For the other boards, I can see that access may be restricted to certain tasks only, so dynamic allocation and switching access rights in context switches would be required.

**Review by evgeniy-paltsev — DISMISSED** [review]

**Review by abrodkin — DISMISSED** [review]

**Review by ruuddw — DISMISSED** [review]

**Comment by evgeniy-paltsev** [comment]
This PR has only arc-related changes and it has all approvals from ARC developers. As we don't have any additional comments we can finally merge this PR.

**Comment by evgeniy-paltsev** [comment]
@fabiobaltieri I guess this can be merged now.

**Review by evgeniy-paltsev — APPROVED** [review]

**Review by ruuddw — APPROVED** [review]

**Review by abrodkin — APPROVED** [review]

**Comment by evgeniy-paltsev** [comment]
@galak may I ask you re-review this? This PR is waiting only your review now.

Thanks!

**Review by galak — APPROVED** [review]

**Pull request merged**
